activation/nipost_test.go

```diff
@@ -807,6 +807,344 @@ func TestNIPSTBuilder_PoetUnstable(t *testing.T) {
 }
 
+// TestNIPoSTBuilder_PoETConfigChange checks if
+// it properly detects added/deleted PoET services and re-registers if needed.
+func TestNIPoSTBuilder_PoETConfigChange(t *testing.T) {
+	t.Parallel()
+
+	sig, err := signing.NewEdSigner()
+	require.NoError(t, err)
+
+	challenge := types.NIPostChallenge{
+		PublishEpoch: postGenesisEpoch + 2,
+	}
+
+	const (
+		poetProverAddr  = "http://localhost:9999"
+		poetProverAddr2 = "http://localhost:9988"
+	)
+
+	t.Run("1 poet deleted BEFORE round started -> continue get proof from known poets", func(t *testing.T) {
+		db := localsql.InMemory()
+
+		require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+		challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+		ctrl := gomock.NewController(t)
+		poetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr)
+		poetProver.EXPECT().Proof(gomock.Any(), "1").AnyTimes().Return(
+			&types.PoetProof{}, []types.Hash32{
+				challengeHash,
+				types.RandomHash(),
+				types.RandomHash(),
+			}, nil,
+		)
+
+		mclock := defaultLayerClockMock(ctrl)
+
+		postClient := NewMockPostClient(ctrl)
+		nonce := types.VRFPostIndex(1)
+		postClient.EXPECT().Proof(gomock.Any(), gomock.Any()).Return(
+			&types.Post{
+				Indices: []byte{1, 2, 3},
+			}, &types.PostInfo{
+				Nonce: &nonce,
+			}, nil,
+		)
+
+		postService := NewMockpostService(ctrl)
+		postService.EXPECT().Client(sig.NodeID()).Return(postClient, nil).AnyTimes()
+
+		// successfully registered to 2 poets
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(10 * time.Second),
+		})
+		require.NoError(t, err)
+
+		err = nipost.AddPoetRegistration(db, sig.NodeID(), nipost.PoETRegistration{
+			ChallengeHash: challengeHash,
+			Address:       poetProverAddr2,
+			RoundID:       "1",
+			RoundEnd:      time.Now().Add(10 * time.Second),
+		})
+
+		// successful post exec
+		nb, err := NewNIPostBuilder(
+			db,
+			postService,
+			zaptest.NewLogger(t),
+			PoetConfig{},
+			mclock,
+			nil,
+			WithPoetServices(poetProver), // add only 1 poet prover
+		)
+		require.NoError(t, err)
+
+		nipost, err := nb.BuildNIPost(context.Background(), sig, challengeHash, &types.NIPostChallenge{PublishEpoch: 7})
+		require.NoError(t, err)
+		require.NotNil(t, nipost)
+	})
+
+	t.Run("1 poet added BEFORE round started -> register to missing poet, get proof from both", func(t *testing.T) {
+		db := localsql.InMemory()
+
+		require.NoError(t, nipost.AddChallenge(db, sig.NodeID(), &challenge))
+		challengeHash := wire.NIPostChallengeToWireV1(&challenge).Hash()
+
+		ctrl := gomock.NewController(t)
+		poetProver := defaultPoetServiceMock(t, ctrl, poetProverAddr)
+		poetProver.EXPECT().Proof(gomock.Any(), "1").AnyTimes().Return(
+			&types.PoetProof{}, []types.Hash32{
+				challengeHash,
+				types.RandomHash(),
+				types.RandomHash(),
+			}, nil,
+		)
```

Comment:
NIT: only the challenge is required in the members list:
```suggestion
		poetProver.EXPECT().Proof(gomock.Any(), "1").AnyTimes().Return(
			&types.PoetProof{}, []types.Hash32{challengeHash}, nil,
		)
```